### ISSUE TITLE
Bonus material: Modes and environments

### ISSUE THREAD

**Issue by susanBuck** [susanBuck/e28-spring20#77]
Hi everyone,

I've just published some bonus information on modes and environments, in response to an earlier thread: [GitGuardian has detected an API key from Google exposed on your GitHub account](https://github.com/susanBuck/e28-spring20/issues/54)

[You can read the notes here...](https://hesweb.dev/e28/notes/vue/environments)

[And here’s an accompanying video...](https://www.youtube.com/embed/7X6sWk_8-YY?rel=0&showinfo=0)

This a bonus/optional topic; you are not required to implement this in your projects.

**Comment by angyts** [comment]
I don't understand, the API key will still be exposed to the browser. So any developer can still pull it out and try to run requests against the firebase database. (including like delete all objects)

It feels like running naked in the snow, leaving a "database" out in full public view...

So now we have to still create authentication, then write more rules about who can access what objects? And each time we create a new collection, to rewrite these rules?

What are some of the security best practices? In a traditional system, for example, only requests made from the server is served by the database, while all other traffic is dropped. Is there a way to lock it down like this? "Only accept requests made from our Vue application?"

**Comment by susanBuck** [comment]
Yes, it's true, the API key will be exposed to the browser. But as noted, the API key (in this particular instance) is just a an identifier for the API - it's not meant to be any sort of security credential.

As far as I know, there's no a way to restrict Firestore access to one particular domain/server. A good explanation on why from [this post](https://stackoverflow.com/a/53425835/59479):

> Firebase security rules aren't able to restrict access to a web domain. In a very general sense, it is not possible, because Firestore is intended to be accessed from mobile clients around the world, using web, Android, and iOS. Android and iOS clients never appear to be coming from some domain. They just directly access the database via the provided client library, or sometimes through the Firestore REST API. Web clients may even spoof their apparent domain (which is only really available by the insecure "Referrer" header in an HTTP request).

Regarding your question:
> So now we have to still create authentication, then write more rules about who can access what objects?

In short, yes. 

Stay tuned for some more bonus info I'll post about authorizing requests. I'll speak to not only Firebase, but other more traditional server API endpoints as well.

**Comment by guyewhite** [comment]
@susanBuck, thank you for this great start to the security side of this. I was able to get all of this set up. I'm like, @punggolzenith now, wanting to ensure that my production server is the only thing that can write to the database -- or how to turn off write capability completely as a protection. In my implementation, my unauthenticated users should be able to read from the database. I'm not in any situation at the moment where I will need to add authentication. Thanks for taking the time to develop these additional materials. I look forward to the next steps.

**Comment by susanBuck** [comment]
FYI Planning on posting this material this coming Monday or Tuesday. : )

**Comment by guyewhite** [comment]
@punggolzenith , I found the following article very helpful: [https://itnext.io/setting-up-cloud-firestore-security-rules-1ad24438805a](https://itnext.io/setting-up-cloud-firestore-security-rules-1ad24438805a)

This one is also good: [https://firebase.google.com/docs/firestore/security/get-started](https://firebase.google.com/docs/firestore/security/get-started)

I was able to restrict to read only for one collection, while keeping the write function enabled for a form submission. Here is how I implemented this....

<img width="1342" alt="Screen Shot 2020-05-02 at 4 41 56 PM" src="https://user-images.githubusercontent.com/46040704/80894889-52c65700-8c94-11ea-8c5a-5482750e3c27.png">

Incidentally, @susanBuck , hopefully you weren't grading my project 2 during the time stamps above ;) as that would surely have made the app "interesting" for a few minutes.

**Comment by susanBuck** [comment]
@guyewhite - I was not. :-)
